addons/mass_mailing/models/mailing_list.py

```diff
@@ -195,68 +195,110 @@ def action_view_contacts_bouncing(self):
         #
-        # The row_column is kind of an occurence counter for the email address.
+        # We retrieve contacts from the lists with distinct groups of identifying data (email, mobile number, ...)
         # Then we create the Many2many relation between the destination list and the contacts
         # while avoiding to insert an existing email address (if the destination is in the source
         # for example)
         self.ensure_one()
         # Put destination is sources lists if not already the case
         src_lists |= self
-        self.env.flush_all()
-        self.env.cr.execute("""
+        self.env['mailing.contact'].flush_model()
+        contact_condition = self._mailing_list_get_contact_condition()
+        select_values = [f"contact.{value}" for value in self._mailing_list_get_select_values()]
+        select_values_string = ', '.join(select_values)
+        select = ', '.join(f"COALESCE({field}, '')" for field in select_values)
+        join_select = ', '.join(f"COALESCE(sub_{field}, '')" for field in select_values) #Need to add coalesce as in postgresql (NULL, 1) = 1
+        insert_query = f"""
             INSERT INTO mailing_contact_list_rel (contact_id, list_id)
-            SELECT st.contact_id AS contact_id, %s AS list_id
-            FROM
-                (
-                SELECT
-                    contact.id AS contact_id,
-                    contact.email AS email,
-                    list.id AS list_id,
-                    row_number() OVER (PARTITION BY email ORDER BY email) AS rn
-                FROM
-                    mailing_contact contact,
-                    mailing_contact_list_rel contact_list_rel,
-                    mailing_list list
-                WHERE contact.id=contact_list_rel.contact_id
-                AND COALESCE(contact_list_rel.opt_out,FALSE) = FALSE
-                AND contact.email_normalized NOT IN (select email from mail_blacklist where active = TRUE)
+            SELECT DISTINCT ON ({select_values_string})
+                contact.id AS contact_id,
+                %(id)s AS list_id
+              FROM mailing_contact contact,
+                  mailing_contact_list_rel contact_list_rel,
+                  mailing_list list
+             WHERE contact.id=contact_list_rel.contact_id
+                AND {contact_condition}
                 AND list.id=contact_list_rel.list_id
-                AND list.id IN %s
-                AND NOT EXISTS
-                    (
-                    SELECT 1
-                    FROM
-                        mailing_contact contact2,
-                        mailing_contact_list_rel contact_list_rel2
-                    WHERE contact2.email = contact.email
-                    AND contact_list_rel2.contact_id = contact2.id
-                    AND contact_list_rel2.list_id = %s
+                AND list.id IN %(ids)s
+                AND ({select}) NOT IN (
+                      SELECT {join_select}
```

Comment:
same here

Reply:
Here I did manage to reduce it to a 'simple' join statement. Had some trouble actually getting it to work, there's something weird going on with UPDATE with JOIN in Postgres so I just put the join in a select...

Tried to do the same for the INSERT statement but it doesn't seem possible since I need to compare to compare the values of the source lists with **every** value of the destination list. Or at least I've spent too long on it already ^^'

-------------

I also parametrized the pre-formated variables that are injected in the statement to make it easier to change the alias name of the contact column (I needed it while tinkering).